add matplotlib, kiwi ports and link to the graalvm blog post (#28)
hpyproject/hpyproject.org: pages/index.rst

- **Much faster** on alternative implementations such as PyPy_, GraalPython_.

- **Universal binaries**: extensions built for the *HPy Universal ABI* can be
  loaded unmodified on CPython, PyPy, GraalPython, etc.

- **A migration path** for mixing legacy C-API calls with HPy API calls. Once

>   all the code is migrated, the extension can be compiled as a universal binary.

Reviewer: ```suggestion
  all the code is migrated, the extension can be compiled as a universal binary that works on any CPython version, PyPy, or GraalPy.
```

(maybe repetition doesn't hurt?)

> - `Debug mode`_: in debug mode, you can easily identify common problems such

Reviewer 2: ```suggestion
- **Debug mode**: in debug mode, you can easily identify common problems such
```
The formatting should be bold here like the others, yes?

Author: I think making it a link is more important, and didn't know how to do both bold and a link.

Reviewer 2: Oh, I see. Maybe we could just add the link next to it in the start of the text? Maybe on the "debug mode" in "in debug mode"?

Author: Good idea, done.

  as memory leaks, invalid lifetime of objects, invalid usage of APIs. Have
  you ever forgot a ``Py_INCREF`` or ``Py_DECREF``? The HPy debug mode can be

>   activated at runtme to detect these mistakes for you on universal binaries.

Reviewer: ```suggestion
  activated at runtime to detect these mistakes for you on universal binaries.
```

.. __: https://hpyproject.org/blog/posts/2022/06/hpy-0.0.4-third-public-release/

The Python/C API is huge and at the moment the most popular functions are

> available. We are experimenting by porting existing extensions to HPy

Author: I seem to recall @antocuni did an analysis of the code in https://github.com/hpyproject/top4000-pypi-packages, can we create a progress report of api-function-name, #times it is used in the packages, checkbox if implemented in HPy

Author: We used to do this for the [micronumpy port](https://buildbot.pypy.org/numpy-status/26e09b343fc3329dab0fbd12e8afb31af4a35808.html) of NumPy in PyPy

Reviewer 2: If we made a list we could also mark ones that we aren't necessarily intending to port.

Reviewer: I made this script some time ago based on one of @antocuni's scripts to analyze top packages with C API usage. It's more focused to find "problematic" usage and give an overview of how many lines of code there would be to touch for an HPy port, but it may be useful to generate a list with: https://gist.github.com/timfel/3fc3b1f339917ff48b69289e79a39f8d

I can take a look tomorrow to adapt it to create such a list.

Author: Cool. Maybe it could be a page we link to here in the future

> The Python/C API is huge and at the moment the most popular functions are
> available. We are experimenting by porting existing extensions to HPy

Reviewer 2: ```suggestion
The Python/C API is huge and at the moment just the most popular functions are
available in HPy. We are have ported moderately sized existing extensions to HPy (e.g. kiwisolver)
and are experimenting with porting larger ones (e.g. numpy).
```

Author: Don't we list that information just below this paragraph? It seems redundant

Author: I think I can rework this section differently

  defines a custom type (similar to ``ndarray`` but with many fewer features of
  course).

- numpy_: one of our ambitious goals is to port numpy to HPy, and to use this
  experience to better understand how to design the API. This builds but does
  not pass the test suite.

- matplotlib_: Since Matplotlib also has a dependency to NumPy, the migration
  to universal mode is not fully finished. HPy provides the legacy compatibility API
  such that we can still call legacy C API functions from HPy and successfully
  run the test suite.

- kiwi-solver_: A dependency of Matplotlib, it has been fully ported to
  universal mode.

Benchmarks and more information about the Matplotlib and kiwi-solver ports can
be found in the `GraalVM blog post`_. The benchmarks show that the HPy ports
have little impact on CPython performance, while enabling close-to-cpython
performance with GraalVM Python for the kiwi-solver complete HPy port.

.. _ultrajson-hpy: https://github.com/hpyproject/ultrajson-hpy
.. _piconumpy: https://github.com/hpyproject/piconumpy

> .. _numpy: https://github.com/hpyproject/numpy-hpy/tree/graal-team/hpy#readme

Author: I think this link is more complete.

  defines a custom type (similar to ``ndarray`` but with many fewer features of
  course).

- numpy_: one of our ambitious goals is to port numpy to HPy, and to use this
  experience to better understand how to design the API. This builds but does
  not pass the test suite.

- matplotlib_: Since Matplotlib also has a dependency to NumPy, the migration
  to universal mode is not fully finished. HPy provides the legacy compatibility API
  such that we can still call legacy C API functions from HPy and successfully
  run the test suite.

- kiwi-solver_: A dependency of Matplotlib, it has been fully ported to
  universal mode.

Benchmarks and more information about the Matplotlib and kiwi-solver ports can
be found in the `GraalVM blog post`_. The benchmarks show that the HPy ports
have little impact on CPython performance, while enabling close-to-cpython
performance with GraalVM Python for the kiwi-solver complete HPy port.

.. _ultrajson-hpy: https://github.com/hpyproject/ultrajson-hpy
.. _piconumpy: https://github.com/hpyproject/piconumpy
.. _numpy: https://github.com/hpyproject/numpy-hpy/tree/graal-team/hpy#readme

> .. _matplotlib: https://github.com/hpyproject/matplotlib-hpy/

Reviewer: ```suggestion
.. _matplotlib: https://github.com/hpyproject/matplotlib-hpy/
.. _kiwisolver: https://github.com/hpyproject/kiwi-hpy
```

Author: Fixed

  experience to better understand how to design the API. This builds but does
  not pass the test suite.

- matplotlib_: Since Matplotlib also has a dependency to NumPy, the migration
  to universal mode is not fully finished. HPy provides the legacy compatibility API
  such that we can still call legacy C API functions from HPy and successfully
  run the test suite.

- kiwi-solver_: A dependency of Matplotlib, it has been fully ported to
  universal mode.

Benchmarks and more information about the Matplotlib and kiwi-solver ports can
be found in the `GraalVM blog post`_. The benchmarks show that the HPy ports
have little impact on CPython performance, while enabling close-to-cpython
performance with GraalVM Python for the kiwi-solver complete HPy port.

.. _ultrajson-hpy: https://github.com/hpyproject/ultrajson-hpy
.. _piconumpy: https://github.com/hpyproject/piconumpy
.. _numpy: https://github.com/hpyproject/numpy-hpy/tree/graal-team/hpy#readme
.. _matplotlib: https://github.com/hpyproject/matplotlib-hpy/
.. _`kiwi-solver`: https://github.com/hpyproject/kiwi-hpy/
.. _`GraalVM blog post`: https://medium.com/graalvm/porting-matplotlib-from-c-api-to-hpy-aa32faa1f0b5

> 

Reviewer 3: Would it be useful to also mention in this page what have been done on Cython? With a HPy backend, it will be possible to compile a lot of projects with universal extensions.

Reviewer: I think the issue is that the Cython work is very early and can only compile a few basic tests right now. We could mention it, though.

Author: I agree. Is there more needed in this PR?

Author: I agree it is very early on, I wouldn't want to mention it yet.

More info
=========

- `Documentation`_

- `HPy blog`_

- `Github repository`_

- #hpy `IRC channel <irc://irc.libera.chat/hpy>`_

- `Mailing list`_

- HPy is pronounced `/h/ - pie` (or using IPA notation: `/eɪtʃ-paɪ/`)

> .. _documentation: https://hpy.readthedocs.io/en/latest/

Reviewer 2: ```suggestion
.. _`Documentation`: https://hpy.readthedocs.io/en/latest/
```
Making formatting like the others again.
